### ISSUE TITLE
Custom File Names

### ISSUE THREAD

**Issue by jonstorer** [achingbrain/mongoose-crate#10]
Hey, would you accept a pull request to allow filename processing? In short a way to configure the file name to be used for the storage strategies?

Specifically, I need a way to put the modelName, fieldName, model's ID, & a time stamp into the file name. The first 3 are for organization & the timestamp is for cache busting.

something like:
`/public/attachments/:modelName/:fieldName/:id-:updated_at-:guid.ext`

**Comment by achingbrain** [comment]
Could you tell me a little more about your use case?

It strikes me you could put a handler behind the URL, use the args to look up the file (e.g. by `:modelName`, `:id`, `:fieldname`) while leaving the cache buster in place but ignored and then stream the file or redirect the user to the s3 URL or whatever.

**Comment by jonstorer** [comment]
I could.

However, the use case I'm solving for is S3 or an S3 like strategy. I'm less worried about LocalFS.

I want to be able to send the asset's url to the client as part of a JSON API's payload. Then have the client to fetch the asset directly. I also want the client to be able to cache assets on device. Which is why I'd like to the url to be representative of the asset it's representing.

I do realize that crate will take the name from the file, and most likely the file will be a tmp file & have a tmp name, however, I can't rely on server to create unique temporary filenames forever. At some point, the file name will repeat and it will overwrite something.

If I could change the storage path at write time, then I could organize the files, include unique information, and cache busting details such as time stamps.

In a real world use case: I had a data snafu where documents lost reference to their assets on S3. I was able to reconnect the asset and document because the storage location included the model name, property name, & doc id.  That was using mongoose-attachment. I'd like to switch to your library because it does not rely on imagemagick. Which means I can have non-image attachments.

I realize, some of this request is vanity. I'd _really_ like to be able to organize the files on S3 is some sane manner. **However**, the part that isn't, is rooted in the real possibility that the files names could collide. And recovering data in case of emergency _is_ possible.

If an asset's path is a combination of model, property, id, etc... the chance of collision is extremely low.

---

About your solution for LocalFS, I'd prefer not to have to retrieve the document from mongo in order to return the asset. I'd rather just expose the asset url to the client. Less work for me.

**Comment by jonstorer** [comment]
@achingbrain also, your library is **so** much cleaner than mongoose-attachments

**Comment by jonstorer** [comment]
@achingbrain ping! I don't mean to bother, but I'd like to get started sooner rather than later. But if you're not going to take the PR, I don't want to do the work.

**Comment by achingbrain** [comment]
I'll merge the PR, a few people have asked for something similar - please do submit it.

**Comment by jonstorer** [comment]
@achingbrain thanks! I'll do this asap (unfortunately it won't be for a couple of days)

**Comment by achingbrain** [comment]
Just a thought, but the storage plugins allow you to override the final path for the file by specifying a `path` function in the plugin options (see https://github.com/achingbrain/mongoose-crate-s3#usage for an example) - would this not accomplish what you are after?

**Comment by jonstorer** [comment]
The _path_ fucntion doesn't have access to the instance of the model. All you have to work with there is the _attachment_ and _this_ which is the scope of the _attach_ function. Believe me, I tried.

**Comment by jonstorer** [comment]
@achingbrain thinking on this again, do you think it makes more sense to pass more info into the `save` method on the storage strategy. That seems to be the only way one could attain the same result.

**Comment by achingbrain** [comment]
Yes, I was coming to the same conclusion - passing the model and the field name into the storage provider's `save` method would probably do it.

**Comment by jonstorer** [comment]
I agree that would allow the strategies to create their own. However, I'm concerned about exposing too much to the strategies. Especially since the full model would be available and someone technically could try to attach an image within the _save_ method (causing a stack trace too deep error) or something else just as awful.

**Comment by jonstorer** [comment]
@achingbrain ping. Are we going to merge this?
